### PR TITLE
[Terraform] Best practices guide

### PR DESCRIPTION
**Pull request by kodster28**
PCX-7622

### CODE REVIEW

File: content/terraform/advanced-topics/best-practices.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+title: Best practices
+---
+
+# Best practices
+
+Though all Terraform deployments are unique, follow these best practices to set yourself up for success.
+
+## Manage Terraform resources in Terraform
+
+Terraform works best when it manages all changes to a resource.
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
Terraform works best when it manages all changes to and the lifecycle of a resource.
```

File: content/terraform/advanced-topics/best-practices.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+After any options on the state file, Terraform attempts to call the read operation and sync the remote into the local state. If there are differences in the local and remote - caused by managing resources outside of Terraform - you will need to delete and recreate the resource in the state file (usually via importing).
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
After any operation on the configuration, Terraform attempts to reconcile the differences by syncing the remote into the local state. If there are differences in the local and remote - caused by managing resources outside of Terraform - you may need to delete and recreate the resource in the state file (usually via importing) as not all resources support in-place updates.
```

File: content/terraform/advanced-topics/best-practices.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+## Directory structure
+
+Cloudflare recommends using a separate based on accounts, zones, and products. This setup lets you have fine-grained owners and scoped Terraform operations to a specific product in a zone. It also allows more closely align owners with Cloduflare's [default roles](/fundamentals/account-and-billing/members/roles/), as well as additional tools like AWS or GCP storage by permissioning separate state files.
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
Cloudflare recommends using a directory structure that relies on a combination of accounts, zones, and products for isolating changes. This setup lets you have fine-grained owners and scoped Terraform operations to a specific product in a zone. It also allows more closely align owners with Cloudflare's [default roles](/fundamentals/account-and-billing/members/roles/), as well as additional tools like AWS or GCP storage by permissioning separate state files.
```

File: content/terraform/advanced-topics/best-practices.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+For Rulesets, you can extend this even further by partitioning at the phase level (WAF, redirects, origin rules).
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
For products that encompass many responsibilities such as Rulesets, you can extend this even further by partitioning at the phase level (WAF, redirects, origin rules).
```

File: content/terraform/advanced-topics/best-practices.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            ├── page_rules.tf
+            ├── provider.tf
+            └── vars.tf
+```
+
+## Avoid modules (or use them sparingly)
+
+Terraform modules are ways of encapsulating multiple resources with logic in an abstracted interface. Consider the example where a module sets up up default load balancer with a pool, some DNS entries, and perhaps a page rule. The end user may use it like this:
+
+```txt
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
```hcl
```

File: content/terraform/advanced-topics/best-practices.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+module "example" "an_example_site" {
+  domain = "example.com"
+  origin_ip = "192.168.0.1"
+}
+```
+
+In reality, however, it will actually be doing this.
+
+```txt
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
```hcl
```

File: content/terraform/advanced-topics/best-practices.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+Terraform modules are ways of encapsulating multiple resources with logic in an abstracted interface. Consider the example where a module sets up up default load balancer with a pool, some DNS entries, and perhaps a page rule. The end user may use it like this:
+
+```txt
+module "example" "an_example_site" {
+  domain = "example.com"
+  origin_ip = "192.168.0.1"
+}
+```
+
+In reality, however, it will actually be doing this.
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
In terms of Terraform resources, however, it will perform this.
```

File: content/terraform/advanced-topics/best-practices.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    minify {
+      html = "off"
+      css  = "on"
+      js   = "on"
+    }
+  }
+}
+```
+
+While convenient, this setup can cause unanticipated issues. If this module is shared and changes internally, the module may have resources out of sync or recreated.
```
Comment (jacobbednarz) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
While convenient, this setup can cause unanticipated issues. If this module is shared and changes internally, the module may have resources out of sync or recreated. 

Using modules also increases the difficulty of debugging or reproducing issues as you must then factor in potential logic bugs outside of Terraform core and the Cloudflare provider.
```

**Review by jacobbednarz — COMMENTED** [review]
thanks for this @kodster28 🚀 couple of wording switcheroos but 👍 overall.

File: content/terraform/advanced-topics/best-practices.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+## Manage Terraform resources in Terraform
+
+Terraform works best when it manages all changes to and the lifecycle of a resource.
+
+After any operation on the configuration, Terraform attempts to reconcile the differences by syncing the remote into the local state. If there are differences in the local and remote - caused by managing resources outside of Terraform - you may need to delete and recreate the resource in the state file (usually via importing) as not all resources support in-place updates.
+
+## Directory structure
+
+Cloudflare recommends using a directory structure that relies on a combination of accounts, zones, and products for isolating changes. This setup lets you have fine-grained owners and scoped Terraform operations to a specific product in a zone. It also allows more closely align owners with Cloudflare's [default roles](/fundamentals/account-and-billing/members/roles/), as well as additional tools like AWS or GCP storage by permissioning separate state files.
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
Cloudflare recommends using a directory structure that relies on a combination of accounts, zones, and products for isolating changes. This setup lets you have fine-grained owners and scoped Terraform operations to a specific product in a zone. It also more closely aligns owners with Cloudflare's [default roles](/fundamentals/account-and-billing/members/roles/), as well as additional tools like AWS or GCP storage by permissioning separate state files.
```

File: content/terraform/advanced-topics/best-practices.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+Terraform modules are ways of encapsulating multiple resources with logic in an abstracted interface. Consider the example where a module sets up up default load balancer with a pool, some DNS entries, and perhaps a page rule. The end user may use it like this:
+
+```hcl
+module "example" "an_example_site" {
+  domain = "example.com"
+  origin_ip = "192.168.0.1"
+}
+```
+
+In terms of Terraform resources, however, it will perform this.
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]: Not entirely sure who `it` in this context is. Who will perform what? Is this something like `In terms of Terraform resources, however, the above example would be translated to:`?

File: content/terraform/advanced-topics/best-practices.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    minify {
+      html = "off"
+      css  = "on"
+      js   = "on"
+    }
+  }
+}
+```
+
+While convenient, this setup can cause unanticipated issues. If this module is shared and changes internally, the module may have resources out of sync or recreated.
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
While convenient, this setup can cause unanticipated issues. If this module is shared and changed internally, the module may have resources out of sync or recreated. 
```

File: content/terraform/advanced-topics/best-practices.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+}
+```
+
+While convenient, this setup can cause unanticipated issues. If this module is shared and changes internally, the module may have resources out of sync or recreated. 
+
+Using modules also increases the difficulty of debugging or reproducing issues as you must then factor in potential logic bugs outside of Terraform core and the Cloudflare provider.
+
+{{<Aside type="warning">}}
+
+This advice also applies to [Terraform dynamic blocks](https://www.terraform.io/language/expressions/dynamic-blocks) that allow you to do logic in your HCL. Since this dynamic blocks are always evaluated, you can get yourself into situations where you have logic bugs in your configuration (and making the end result unreproducible).
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
This advice also applies to [Terraform dynamic blocks](https://www.terraform.io/language/expressions/dynamic-blocks) that allow you to do logic in your HCL. Since these dynamic blocks are always evaluated, you can get yourself into situations where you have logic bugs in your configuration (and making the end result unreproducible).
```

File: content/terraform/advanced-topics/best-practices.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+To safely manage separate environments (staging, QA, UAT, production), use separate Cloudflare accounts with separate domains (such as `example.com` and `example-staging.com`).
+
+This is because some products defined at the account level are shared (such as Load Balancer monitors and pools) and you cannot make an isolated change to them if they are in the same account. Using separate accounts is also beneficial if you intend to test things like DNSSEC, which may affect your entire domain if configured incorrectly.
+
+To minimize drift, use Terraform and a CI/CD pipeline that runs across both domains to keep them in sync as needed.
+
+## Store credentials safely
+
+We recommend not storing Cloudflare credentials as plaintext.
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
We do not recommend storing Cloudflare credentials as plaintext.
```

Comment (marciocloudflare) [reply content/terraform/advanced-topics/best-practices.md]: I think it's easier to understand this way

File: content/terraform/advanced-topics/best-practices.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+Locally, you can use a third-party tool like [cf-vault](https://github.com/jacobbednarz/cf-vault/).
```
Comment (marciocloudflare) [inline content/terraform/advanced-topics/best-practices.md]:
```suggestion
Locally, you can use a third-party tool like [cf-vault](https://github.com/jacobbednarz/cf-vault/) to store your Cloudflare credentials.
```

Comment (marciocloudflare) [reply content/terraform/advanced-topics/best-practices.md]: I think the sentence benefits from something here. Wdyt?

**Review by marciocloudflare — APPROVED** [review]

File: content/terraform/advanced-topics/best-practices.md (hunk 10 above)
Comment (kodster28) [reply content/terraform/advanced-topics/best-practices.md]: Good callout. I'll reword.

File: content/terraform/advanced-topics/best-practices.md (hunk 11 above)
Comment (kodster28) [reply content/terraform/advanced-topics/best-practices.md]: Might slightly reword this to, `If this module is shared and then changes internally`.

It's one of those rare things where we're dealing with multiple tenses.
